<?xml version="1.1"?>
<launch>
    <!-- Argument for the world frame ID -->
    <arg name="world_frame_id" default="world" /> 
    <arg name="setting_file_path" default="/home/gyanig/catkin_ws/src/relaxed_ik_ros1/relaxed_ik_core/configs/settings.yaml" />
    <param name="setting_file_path" value="$(arg setting_file_path)" />
    <param name="/right_cam/realsense2_camera/initial_reset" value="true" />
    
    <arg name="serial_no"             default="838212072405"/>
    <arg name="json_file_path"        default=""/>
    <arg name="camera"                default="right_cam"/>
    

    <node pkg="joy" type="joy_node" name="xbox" output="screen">
        <param name="dev" value="/dev/input/js0" />
    </node>

    <group ns="$(arg camera)">
        <include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
        <arg name="serial_no"         value="$(arg serial_no)"/>
        <arg name="json_file_path"    value="$(arg json_file_path)"/>
        <arg name="depth_width"       value="640"/>
        <arg name="depth_height"      value="480"/>
        <arg name="depth_fps"         value="30"/>
        <arg name="color_width"       value="640"/>
        <arg name="color_height"      value="480"/>
        <arg name="color_fps"         value="30"/>
        <arg name="enable_depth"      value="true"/>
        <arg name="enable_color"      value="true"/>
        <arg name="enable_infra1"     value="false"/>
        <arg name="enable_infra2"     value="false"/>
        <arg name="enable_fisheye"    value="false"/>
        <arg name="enable_gyro"       value="false"/>
        <arg name="enable_accel"      value="false"/>
        <arg name="enable_pointcloud" value="true"/>
        <arg name="enable_sync"       value="true"/>
        <arg name="align_depth"       value="true"/>
        <arg name="tf_prefix"         value="$(arg camera)"/>
    </include>
    </group>
            <!-- <arg name="filters"           value="decimation,disparity,spatial,temporal,hole_filling, colorizer"/> -->

    <!-- <node name="rviz" pkg="rviz" type="rviz" args="-d $(find realsense2_camera)/rviz/pointcloud.rviz" required="true" /> -->
    
<!--     
    <node pkg="sawyer_teleop" type="hand_tracker.py" name="hand_tracker" output="screen">
        <param name="image_topic" value="/cropped_table_view/color/image_raw"/>
        <param name="show_gui"   value="true"/>
        <param name="use_mediapipe" value="true"/>
    </node> -->

    <!-- Sawyer Robot Transform -->
    
    <node pkg="tf" type="static_transform_publisher" name="world_to_sawyer_tf"
          args="0.0 0.0 0.0 0 0 0 $(arg world_frame_id) sawyer_base_link 100" />

<!-- 
    <node pkg="tf" type="static_transform_publisher" name="world_to_right_cam_tf"
          args="0.71249 0.9124 0.96249 271 53 12 $(arg world_frame_id) right_cam_link 100" />
-->
    <!-- <node pkg="tf" type="static_transform_publisher" name="world_to_right_cam_color_optical_frame_tf"
          args="0.71249 0.9124 0.96249 271 53 12 $(arg world_frame_id) right_cam_color_optical_frame 100" />  -->


    <node pkg="sawyer_teleop" type="joint_velocity_controller.py" name="joint_velocity_controller" output="screen"/> 
    <node pkg="relaxed_ik_ros1" type="relaxed_ik_rust.py" name="relaxed_ik_rust" output="screen" >
        <param name="use_visualization" value="false" />
    </node>

    <!-- <node pkg="sawyer_teleop" type="joystick_input.py" name="joystick_teleop" output="screen" args="-j xbox -l right"/>  -->
    <node pkg="relaxed_ik_ros1" type="keyboard_input.py" name="keyboard_teleop" output="screen" /> 

    <node pkg="sawyer_teleop" type="stitch_object_detection_vt.py" name="stitch_object_detection" output="screen"> 
        <param name="image_topic"         value="/right_cam/color/image_raw"/>
        <param name="depth_topic"         value="/right_cam/aligned_depth_to_color/image_raw"/>
        <param name="cam_info_topic"      value="/right_cam/color/camera_info"/>
        <param name="color_optical_frame" value="right_cam_color_optical_frame"/>
        <param name="base_frame"          value="world"/>
        <param name="model"         value="yolov8m.pt"/>       
        <param name="conf_thres"    value="0.4"/>
        <param name="iou_thres"     value="0.3"/>
        <param name="classes"       value="fan, headphone, bowl, remote, bottle, cup, fruits"/> 
        <param name="patch_radius_px" value="7"/>   
        <param name="min_depth_m"     value="0.15"/>
        <param name="max_depth_m"     value="5.0"/>

        <param name="show_gui" value="true"/>
    </node> 

    <node pkg="sawyer_teleop" type="intent_inference.py" name="intent_inference" output="screen">
        <param name="end_effector_topic" value="/robot/limb/right/endpoint_state"/> 
        <param name="detections_topic" value="/stitch_object_detection/detections"/>
        <param name="stationary_speed_mps" value="0.02"/>
        <param name="base_frame" value="world"/>
    </node>

    <node pkg="sawyer_teleop" type="intent_viz_tracker.py" name="intent_viz" output="screen">
        <param name="det_topic"  value="/stitch_object_detection/detections"/>
        <param name="prob_topic" value="/intent_inference/distribution"/>
        <param name="img_topic"  value="/stitch_object_detection/annotated_image"/>
        <param name="show_combined" value="true"/>
    </node>
     <!-- Trajectory planner -->
    <!-- <node pkg="sawyer_teleop" type="tabletop_trajectory_following.py" name="yolo_waypoint_planner" output="screen"/> -->

    <!-- RViz node with a custom configuration file -->
    <node pkg="rviz" type="rviz" name="rviz_viewer_scene" args="-d $(find sawyer_teleop)/config/camera_robot.rviz" output="screen" />


</launch>